core/trino-main/src/main/java/io/trino/server/NodeStateManager.java

```diff
@@ -60,7 +61,37 @@ public class NodeStateManager
     private final Duration gracePeriod;
 
     private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor(threadsNamed("drain-handler-%s"));
-    private final AtomicReference<NodeState> nodeState = new AtomicReference<>(ACTIVE);
+    private final AtomicReference<VersionedState> nodeState = new AtomicReference<>(new VersionedState(ACTIVE, 0));
+
+    private record VersionedState(NodeState state, long version)
+    {
+        private static final AtomicLong versionProvider = new AtomicLong(0);
```

Comment:
put it in `NodeStateManager` and make non-static

Reply:
Also it does not need to be atomic as it is accessed from synchronized blocks.

Reply:
Now it is static, so a single instance for a VersionedState class in a single classloader, when I move to NodeStateManager and remove `static` it will be scoped within a `NodeStateManager` instance. Ok, probably makes sense. 

But then I also need to provide version from nodestate manager to all transition methods (`toActive`, `toDraining` etc.), alternatively I need to move these methods to NodeStateManager. Do you think this is cleaner design?

Reply:
Yeah - this is good point. I added a commit with proposed code change
tl;dr; VersinedState knows about NSM

Reply:
Nice proposition, I have actually applied this change.